### PR TITLE
Feature/test framework enhancments

### PR DESCRIPTION
**Pull request by mpustelnik**
TODO: 
- [x] write test samples to cover test scenarios with added features
- [x] consider creating SendsEvent extension method for User **at the moment we don't need this feature as it can be done by SendsActivity. We can implement it in the future**

### CODE REVIEW

File: src/V4/Objectivity.Bot.Tests.Stories.Xunit.V4/Extensions/BotStoryRecorderExtensions.cs [hunk 1]
```diff
@@ -29,6 +31,46 @@ public static IStoryRecorder<IMessageActivity> GivesChoice(this IBotRecorder<IMe
...
+        /// </summary>
+        /// <param name="recorder">Bot recorder.</param>
+        /// <param name="name">Event name.</param>
+        /// <param name="valuePredicate">Predicate used for event value assertion.</param>
+        /// <typeparam name="T">Value type.</typeparam>
+        /// <returns>Story recorder.</returns>
+        public static IStoryRecorder<IMessageActivity> SendsEvent<T>(
+            this IBotRecorder<IMessageActivity> recorder,
+            string name,
+            Func<T, bool> valuePredicate)
```
Comment (piotrzajac) [inline src/V4/Objectivity.Bot.Tests.Stories.Xunit.V4/Extensions/BotStoryRecorderExtensions.cs]: Consider having just a eventActivityPredicate:
```csharp
Func<IEventActivity, bool> eventActivityPredicate
```
That way we can achieve pretty much the same:
```csharp
return recorder.SendsActivity(activity => activity != null
                                          && activity.Type == ActivityTypes.Event
                                          && eventActivityPredicate(activity.AsEventActivity()));
```

File: src/V4/Objectivity.Bot.Tests.Stories.Xunit.V4/Extensions/BotStoryRecorderExtensions.cs [hunk 2]
```diff
@@ -29,6 +31,46 @@ public static IStoryRecorder<IMessageActivity> GivesChoice(this IBotRecorder<IMe
...
+        /// <param name="recorder">Bot recorder.</param>
+        /// <param name="name">Event name.</param>
+        /// <returns>Story recorder.</returns>
+        public static IStoryRecorder<IMessageActivity> SendsEvent(
+            this IBotRecorder<IMessageActivity> recorder,
+            string name)
+        {
+            return recorder.SendsActivity(activity => activity != null
+                                                      && activity.Type == ActivityTypes.Event
+                                                      && activity.AsEventActivity()?.Name == name);
```
Comment (piotrzajac) [inline src/V4/Objectivity.Bot.Tests.Stories.Xunit.V4/Extensions/BotStoryRecorderExtensions.cs]: Just a minor comment. You validate if event activity is not null here but not in TestAuthPrompt.OnRecognizeAsync method. What's the reason?

File: src/Common/Objectivity.Bot.Tests.Stories/Core/CustomActivityBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+        private readonly T activity;
+
+        public CustomActivityBuilder(T activity)
+        {
+            this.activity = activity;
+        }
+
+        public T Build(IStoryFrame<T> frame)
+        {
+            return this.activity;
```
[the remark this answers is not in the document]
Comment (mpustelnik) [reply src/Common/Objectivity.Bot.Tests.Stories/Core/CustomActivityBuilder.cs]: Frame parameter is forced by common interface implemented by this class. In this case this is just a wrapper for custom activity prepared by the user

File: src/V4/Objectivity.Bot.Tests.Stories.Xunit.V4.DemoBot/Dialogs/Prompts/TestAuthPrompt.cs [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        }
+
+        protected override Task<PromptRecognizerResult<Activity>> OnRecognizeAsync(
+            ITurnContext turnContext,
+            IDictionary<string, object> state,
+            PromptOptions options,
+            CancellationToken cancellationToken = default(CancellationToken))
+        {
+            var result = new PromptRecognizerResult<Activity>();
+            var activity = turnContext.Activity;
```
[the remark this answers is not in the document]
Comment (mpustelnik) [reply src/V4/Objectivity.Bot.Tests.Stories.Xunit.V4.DemoBot/Dialogs/Prompts/TestAuthPrompt.cs]: This is just a sample. These asserts would add unnecessary complexity to the code.

File: src/V4/Objectivity.Bot.Tests.Stories.Xunit.V4.DemoBot/Dialogs/Prompts/TestAuthPrompt.cs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    using System.Collections.Generic;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Bot.Builder;
+    using Microsoft.Bot.Builder.Dialogs;
+    using Microsoft.Bot.Schema;
+
+    public class TestAuthPrompt : ActivityPrompt
+    {
+        public const string AuthRequestEventName = "tokens/request";
```
[the remark this answers is not in the document]
Comment (mpustelnik) [reply src/V4/Objectivity.Bot.Tests.Stories.Xunit.V4.DemoBot/Dialogs/Prompts/TestAuthPrompt.cs]: This is just a sample. In this case keeping these variables inside this class makes the code easier to understand.

File: src/V4/Objectivity.Bot.Tests.Stories.Xunit.V4/Extensions/BotStoryRecorderExtensions.cs (hunk 2 above)
Comment (mpustelnik) [reply src/V4/Objectivity.Bot.Tests.Stories.Xunit.V4/Extensions/BotStoryRecorderExtensions.cs]: The reason is that TestAuthPrompt is just a sample code. This code is a part of the library so we have to make a null check.

File: src/V4/Objectivity.Bot.Tests.Stories.Xunit.V4.DemoBot/Dialogs/Prompts/TestAuthPrompt.cs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+            var activity = turnContext.Activity;
+
+            if (activity.Type == ActivityTypes.Event)
+            {
+                var ev = activity.AsEventActivity();
+
+                if (ev.Name == AuthResponseEventName)
+                {
+                    result.Succeeded = true;
+                    result.Value = turnContext.Activity;
```
[the remark this answers is not in the document]
Comment (mpustelnik) [reply src/V4/Objectivity.Bot.Tests.Stories.Xunit.V4.DemoBot/Dialogs/Prompts/TestAuthPrompt.cs]: Done. Used previously assigned activity variable

File: src/V4/Objectivity.Bot.Tests.Stories.Xunit.V4.DemoBot/Dialogs/TestEventDialog.cs [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+            authEventReply.Name = TestAuthPrompt.AuthRequestEventName;
+
+            await stepContext.Context.SendActivityAsync(authEventReply, cancellationToken);
+
+            return await stepContext.PromptAsync(nameof(AuthDialogId), new PromptOptions(), cancellationToken);
+        }
+
+        private async Task<DialogTurnResult> ShowAuthorizedAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
+        {
+            await stepContext.Context.SendActivityAsync("You're authorized", cancellationToken: cancellationToken);
```
[the remark this answers is not in the document]
Comment (mpustelnik) [reply src/V4/Objectivity.Bot.Tests.Stories.Xunit.V4.DemoBot/Dialogs/TestEventDialog.cs]: Added named argument.

File: src/V4/Objectivity.Bot.Tests.Stories.Xunit.V4.Sample/Dialog/TestEventDialogTests.cs [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    using Objectivity.Bot.Tests.Stories.Core;
+    using Objectivity.Bot.Tests.Stories.Xunit.V4.Extensions;
+
+    public class TestEventDialogTests : DialogTestBase<TestEventDialog>
+    {
+        [Fact]
+        public async Task GivenDialogStarts_PlayStoryIsCalled_MustReturnAuthorizationRequestEvent()
+        {
+            var story = this.Record
+                .Bot.SendsEvent("tokens/request")
```
[the remark this answers is not in the document]
Comment (mpustelnik) [reply src/V4/Objectivity.Bot.Tests.Stories.Xunit.V4.Sample/Dialog/TestEventDialogTests.cs]: Done.

File: src/V4/Objectivity.Bot.Tests.Stories.Xunit.V4.DemoBot/Dialogs/TestEventDialog.cs [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+            var activity = pc.Recognized.Value;
+
+            if (activity == null || activity.Type != ActivityTypes.Event)
+            {
+                await pc.Context.SendActivityAsync("You're unauthorized", cancellationToken: cancellationToken);
+
+                return false;
+            }
+
+            return true;
```
[the remark this answers is not in the document]
Comment (mpustelnik) [reply src/V4/Objectivity.Bot.Tests.Stories.Xunit.V4.DemoBot/Dialogs/TestEventDialog.cs]: Changed this part according to your suggestion.

File: src/V4/Objectivity.Bot.Tests.Stories.Xunit.V4/Extensions/BotStoryRecorderExtensions.cs [hunk 10]
```diff
@@ -29,6 +31,46 @@ public static IStoryRecorder<IMessageActivity> GivesChoice(this IBotRecorder<IMe
                        && actions.Count == options.Length
                        && options.All(x => actions.Any(a => a.Title == x));
             });
+        }
+
+        /// <summary>
+        /// Verifies if bot sent an event.
```
[the remark this answers is not in the document]
Comment (mpustelnik) [reply src/V4/Objectivity.Bot.Tests.Stories.Xunit.V4/Extensions/BotStoryRecorderExtensions.cs]: Added an information.

File: src/V4/Objectivity.Bot.Tests.Stories.Xunit.V4/Extensions/BotStoryRecorderExtensions.cs (hunk 1 above)
Comment (mpustelnik) [reply src/V4/Objectivity.Bot.Tests.Stories.Xunit.V4/Extensions/BotStoryRecorderExtensions.cs]: Good suggestion. Thank you. Changed this method overload to accept event predicate only.

File: src/V4/Objectivity.Bot.Tests.Stories.Xunit.V4.DemoBot/Dialogs/TestEventDialog.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (piotrzajac) [reply src/V4/Objectivity.Bot.Tests.Stories.Xunit.V4.DemoBot/Dialogs/TestEventDialog.cs]: But not in the line #50.

File: src/V4/Objectivity.Bot.Tests.Stories.Xunit.V4/Extensions/BotStoryRecorderExtensions.cs [hunk 11]
```diff
@@ -29,6 +31,43 @@ public static IStoryRecorder<IMessageActivity> GivesChoice(this IBotRecorder<IMe
...
+
+        /// <summary>
+        /// Verifies if bot sent an event.
+        /// <remarks>
+        /// Assertion is based on activity type, name and value predicate. Name verification is case sensitive.
+        /// </remarks>
+        /// </summary>
+        /// <param name="recorder">Bot recorder.</param>
+        /// <param name="eventPredicate">Predicate used for event assertion.</param>
+        /// <typeparam name="T">Value type.</typeparam>
```
Comment (piotrzajac) [inline src/V4/Objectivity.Bot.Tests.Stories.Xunit.V4/Extensions/BotStoryRecorderExtensions.cs]: Do we need this type at all in this case?

Comment (mpustelnik) [reply src/V4/Objectivity.Bot.Tests.Stories.Xunit.V4/Extensions/BotStoryRecorderExtensions.cs]: Of course not :) thanks for catching this one

File: src/V4/Objectivity.Bot.Tests.Stories.Xunit.V4/Extensions/BotStoryRecorderExtensions.cs [hunk 12]
```diff
@@ -29,6 +31,42 @@ public static IStoryRecorder<IMessageActivity> GivesChoice(this IBotRecorder<IMe
...
+        {
+            return recorder.SendsActivity(activity => activity != null
+                                                      && activity.Type == ActivityTypes.Event
+                                                      && activity.AsEventActivity()?.Name == name);
+        }
+
+        /// <summary>
+        /// Verifies if bot sent an event.
+        /// <remarks>
+        /// Assertion is based on activity type, name and value predicate. Name verification is case sensitive.
```
Comment (piotrzajac) [inline src/V4/Objectivity.Bot.Tests.Stories.Xunit.V4/Extensions/BotStoryRecorderExtensions.cs]: You have just improved the code this the comment should follow the update.

Comment (mpustelnik) [reply src/V4/Objectivity.Bot.Tests.Stories.Xunit.V4/Extensions/BotStoryRecorderExtensions.cs]: Fixed

**Pull request merged**